### ISSUE TITLE
WebAuthN doesn't work with example nginx config

### ISSUE THREAD

**Issue by vaminakov** [zitadel/zitadel#7251]
**Edit by @muhlemmer** 

TL;DR: The example [Nginx config](https://zitadel.com/docs/self-hosting/manage/reverseproxy/nginx) rewrites the request `Host` by appending a port. This breaks WebAuthN registration, as the signed origin does not match the actual request origin. This problem seems to present itself using the standard 443 port, instead of the custom 8080 port that the guide assumes.

- If there is a custom port deployment, the port _should_ be set in the `Host` header
- If there is a standard port deployment, the port _should not_ be set in the `Host` header.

Question: why rewrite the host header at all? Why not forward the original Host?

------

**Original issue**

### Preflight Checklist

- [X] I could not find a solution in the documentation, the existing issues or discussions
- [ ] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Environment

ZITADEL Cloud

### Version

2.43.5

### Database

PostgreSQL

### Database Version

16.1

### Describe the problem caused by this bug

There are no possibility to register any webauthn/passkey device in fresh install.

### To reproduce

1.     Create a new instance
2.     Login to the instance with username + password
3.     Go to User setting > Password and security
4.     Click "add method" on passwordless or "add factor" on multifactor
5.     Enter any device name
6.     Get popup error "Error on create credentials" while in zitadel admin or "Errors.User.WebAuthN.CreateCredentialFailed" while in 2fa registration by link.

**Comment by hifabienne** [comment]
Hei @vaminakov 
I quickly checked this myself and didn't have any problems.
Can you please add some more information about your environment?
Which os do you have? What browser do you use?

And I just saw that you say you are on our cloud but you also added db version, so it looks like you are running your own system? So just want to make sure I understand right were you are at

**Comment by vaminakov** [comment]
@hifabienne hello! Sorry for delay in response.
I have own installation (cloud in description is mistake).
I'm tried to debug this problem in OS ArchLinux, Ubuntu 23.10. Browser - latest Chrome and Firefox.
Anywhere problem is reproducing.

As reverse-proxy i'm using nginx with config:
```
location / { 
modsecurity off; 
grpc_pass grpc://localhost:8098; 
grpc_set_header Host $host:$server_port; 
}
```

**Comment by livio-a** [comment]
Hey @vaminakov 

Does your reverse proxy handle TLS? Resp. can you share your configuration of zitadel (ExternalDomain, ExternalPort, ExternalSecure, Port)?

**Comment by vaminakov** [comment]
@livio-a hello! Yes, there're full config:
nginx:

```
server {
server_name zitadel.mydomain;
listen *:443 ssl;
listen [::]:443 ssl;
listen *:443 quic;
listen [::]:443 quic;
ssl_certificate fullchain.pem;
ssl_certificate_key privkey.pem;
ssl_stapling on;
ssl_stapling_verify on;
ssl_stapling_file ocsp.der;
add_header Strict-Transport-Security "max-age=63072000; includeSubDomains; preload";
add_header Public-Key-Pins 'pin-sha256="mykey"; max-age=63072000';
add_header Alt-Svc 'h3=":443"; ma=86400';
add_header X-protocol $server_protocol always;
location / { 
    grpc_pass grpc://localhost:8098; 
    grpc_set_header Host $host:$server_port; 
               }
}
```
Zitadel:

```
ExternalSecure: true
ExternalDomain: zitadel.mydomain
ExternalPort: 443
Port: 8098
WebAuthNName: zitadel.mydomain
Database:
    postgres:
        Host: localhost
        Port: 5432
        Database: zitadel
        MaxOpenConns: 25
        MaxConnLifetime: 1h
        MaxConnIdleTime: 5m
        Options:
        User:
            Username: zitadel
            Password: secretpass
            SSL:
                Mode: disable
                RootCert:
                Cert:
                Key:
        Admin:
            Username: postgres
            Password: secretpass
            SSL:
                Mode: disable
                RootCert:
                Cert:
                Key:
LogStore:
  Access:
    Stdout:
      Enabled: true
  Execution:
    Stdout:
      Enabled: true
Log:
  Level: info
  Formatter:
    Format: text
Metrics:
  Type: none
```

Systemd unit:

```
# /etc/systemd/system/zitadel.service
[Unit]
Description=Zitadel IAM
After=postgresql.service

[Service]
RestartSec=2s
Type=simple
User=zitadel
WorkingDirectory=/var/lib/zitadel
Environment=USER=zitadel HOME=/var/lib/zitadel
ExecStart=/usr/bin/zitadel start --config /etc/zitadel/config.yaml --masterkeyFile /etc/zitadel/master.key --tlsMode external
ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=on-failure

[Install]
WantedBy=multi-user.target
```

**Comment by livio-a** [comment]
The config seems ok. Can you start zitadel with log level debug? 
You should get some more info then: https://github.com/zitadel/zitadel/blob/f680dd934df9c1ff2b3b91b4565997cb2c2db6a2/internal/webauthn/webauthn.go#L119

might be related to #2910

**Comment by vaminakov** [comment]
@livio-a 
```
level=debug msg="webauthn credential could not be created" caller="/home/runner/work/zitadel/zitadel/internal/webauthn/webauthn.go:119" error="Error validating origin: Expected Values: [https://zitadel.mydomain:443], Received: https://zitadel.mydomain"
```

**Comment by muhlemmer** [comment]
>```
>level=debug msg="webauthn credential could not be created" caller="/home/runner/work/zitadel/zitadel/internal/webauthn/webauthn.go:119" error="Error validating origin: Expected Values: [https://zitadel.mydomain:443], Received: https://zitadel.mydomain"
>```

That explains it, the Host in the request is re-written with the port number appended. `https://zitadel.mydomain:443` != `https://zitadel.mydomain`. We take and configure the `RPOrigins` from the request `Host` header here:

https://github.com/zitadel/zitadel/blob/3bbf38acc5e1f3d9e7432efaa852a45a1f280972/internal/webauthn/webauthn.go#L197-L204

That's because zitadel is able to host multiple instance with their own domain, so we need to populate that value dynamically.

The WebAuthN library compares the strings of the config's `RPOrigins` against the Origin in the signed response from the browser. It the browser thinks it's operating with origin `https://zitadel.domain` and the origin from the `Host` is rewritten, it won't work.

From the nginx config posted above:

> grpc_set_header Host $host:$server_port; 

I suspect that is the reason why the `Host` header that zitadel receives has the additional port number.

**Comment by vaminakov** [comment]
@muhlemmer so maybe possible to exclude external server port from checking in this case? Or make it identical in variable?

**Comment by muhlemmer** [comment]
No, that would break deployments that run on a non-standard ports, like https://example.com:8080. Why not drop the port from your proxy config?

**Comment by vaminakov** [comment]
@muhlemmer You're right! I've dropped port from proxy config and all working now. 
Perhaps this should be reflected in the documentation?

**Comment by muhlemmer** [comment]
Yes, we would need to update https://zitadel.com/docs/self-hosting/manage/reverseproxy/nginx to reflect this properly. Although that  documentation assumes a custom port (8080) and in that case it seems to make sense to append the port. I will forward this issue to devops so they can double check which combinations work.

Or, if you you like you can submit a PR to update the docs, but please make sure that both standard and custom port deployments work as expexted.

**Comment by muhlemmer** [comment]
CC @eliobischof

**Comment by vaminakov** [comment]
I'll try to do this.
Not entirely related to the topic, but it would be nice (for the devops block) to clarify which locations are required for zitadel to work. This is needed to block on a conditional WAF location /
Now it seems to
```
location ~ ^/(?:ui|oauth|oidc|.well-known|zitadel.*|assets/v1) { grpc_pass grpc://zitadel; grpc_set_header Host $host; }
location / { return 444; }
```

**Comment by eliobischof** [comment]
Not sure how easy it is to achieve, but the more sustainable solution would be if ZITADEL wouldn't care if default ports are given or not.

**Comment by muhlemmer** [comment]
> Not sure how easy it is to achieve

It's a documentation update, remove the port from the example. What is not easy?

>  but the more sustainable solution would be if ZITADEL wouldn't care if default ports are given or not.

We and libraries do string based compare on multiple occasions. For example

1. `https://example.com != https://example.com/` 
2. `https://example.com:443 != https://example.com`

This applies to WebauthN (library not under our controll), [OIDC issuer](https://github.com/zitadel/zitadel/issues/7804) and origin parsing, which must be direct string compare according OIDC standard.

**Comment by eliobischof** [comment]
This was done with a2f2ffb737978795125211371b53f15c01bb682f
